### DISCUSSION TITLE
a few questions about optimizing tiffsave

### DISCUSSION THREAD

**Discussion by zhusihan-python** [libvips/libvips#3449] [Help wanted]
hello, i stitched some images tiles by Image.merge, and noticed that the pipeline will not start before tiffsave is call,  i have a few questions about how to optimize tiffsave : 
1. is there a way to trigger the pipeline in advance, so it will calculate right after i feed it a tile by merge, so the total tiffsave time will be less

2. my cpu is 16 physical core 32 logical core, when set VIPS_CONCURRENCY to 4, 16, 32, the cpu usages keeps less than 10% most of the time, the memory usage peak is around 25G, 39G remaining (VIPS_CONCURRENCY=64) 38G used, 26G remaining (VIPS_CONCURRENCY=32), and the tiffsave time almost the same, even longer when the concurrency bigger than cpu core numbers, is there something else restrict the tiffsave parallel

3. the background=list[float] param seems not working, the color near borders is always black, no matter i set background=[255.0, 255.0, 255.0] or background=[0.0, 0.0, 0.0]
![image](https://user-images.githubusercontent.com/23504604/232703772-fc12b7ce-8231-4f71-b94f-eae4c4c0fc1a.png)

**Comment by jcupitt** [comment]
Hi, tiffsave is currently not threaded, so assuming you are using tiles jpeg tiff, your save speed is limited by the single-threaded jpeg compress.

We added threading to tiffload in 8.14 and there's talk of adding threading to tiffsave during 8.15, but no one's done it yet. 

There are some notes and benchmarks here:

https://www.libvips.org/2022/12/22/What's-new-in-8.14.html#tiff

**Reply by zhusihan-python** [reply]
thanks for the quick reply. i'm using tiles jpeg tiff, and i build libvips 8.14 with jpeg turbo, it will be also single-threaded jpeg right though it compress faster.

hope i can try the multi threading tiffsave in 8.15.

and about the third question, what should i pass to tiffsave background param if i want to set tiff file background to white color

**Reply by zhusihan-python** [reply]
hello @jcupitt  john, sorry to bother, i wonder if there is any plan about adding threading to tiffsave in 8.15

**Reply by jcupitt** [reply]
I'm trapped in another project until July. I might have time to look at this then.
